src/cursor/abstract_cursor.ts

```diff
@@ -757,18 +719,71 @@ function cursorIsDead(cursor: AbstractCursor): boolean {
   return !!cursorId && cursorId.isZero();
 }
 
-function cleanupCursor(cursor: AbstractCursor, callback: Callback): void {
-  if (cursor[kDocuments].length === 0) {
-    cursor[kClosed] = true;
-    cursor.emit(AbstractCursor.CLOSE);
+function cleanupCursor(
+  cursor: AbstractCursor,
+  options: { error?: AnyError | undefined; needsToEmitClosed?: boolean } | undefined,
+  callback: Callback
+): void {
+  const cursorId = cursor[kId];
+  const cursorNs = cursor[kNamespace];
+  const server = cursor[kServer];
+  const session = cursor[kSession];
+  const error = options?.error;
+  const needsToEmitClosed =
+    options?.needsToEmitClosed ||
+    (options?.needsToEmitClosed == null && cursor[kDocuments].length === 0);
+
+  if (error) {
+    if (cursor.loadBalanced && error instanceof MongoNetworkError) {
+      return completeCleanup();
+    }
   }
 
-  const session = cursor[kSession];
-  if (session && session.owner === cursor) {
-    session.endSession(callback);
-  } else {
-    callback();
+  if (cursorId == null || server == null || cursorId.isZero() || cursorNs == null) {
+    if (needsToEmitClosed) {
+      cursor[kClosed] = true;
+      cursor[kId] = Long.ZERO;
+      cursor.emit(AbstractCursor.CLOSE);
+    }
+
+    if (session) {
```

Comment:
I've wrapped this all into `completeCleanup` now, which also fixed the cursor's close event from getting emitted multiple times. Ever I left a todo to rename an option name to the cursor since we've already released 4.0.